File: frontend/src/api/images.ts
import { api } from "./http";

export type LabeledImage = {
  id: string;
  image_url: string;
  marbling_class: string;
  created_at: string;
};

export async function uploadLabeledImage(
  file: File,
  marbling_class: string,
): Promise<LabeledImage> {
  const fd = new FormData();
  fd.append("file", file);
  fd.append("marbling_class", marbling_class);

  return api<LabeledImage>("/api/images", { method: "POST", body: fd });
}

File: back/app/routers/images.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status

from app.core.auth_deps import get_current_user
from app.core.config import settings
from app.core.constants import NINE_CLASSES
from app.core.storage import save_upload_to_media
from app.models.game_image import GameImageOut
from app.services.game_images_service import GameImagesService

router = APIRouter(prefix="/images", tags=["images"])


@router.post("", response_model=GameImageOut, status_code=status.HTTP_201_CREATED)
async def upload_labeled_image(
    file: UploadFile = File(...),
    marbling_class: str = Form(...),
    current_user: dict = Depends(get_current_user),
) -> GameImageOut:
    
    if marbling_class not in NINE_CLASSES:
        raise HTTPException(
            status_code=422,
            detail=f"marbling class should be one of : {', '.join(NINE_CLASSES)}",
        )

    saved = await save_upload_to_media(file)
    image_url = settings.build_public_url(saved["rel_path"])

    created = await GameImagesService.create(
        image_path=saved["rel_path"],
        image_url=image_url,
        marbling_class=marbling_class,
    )
    return created
